### PR TITLE
Consul Pillar: select DC based on pillarenv

### PR DESCRIPTION
**Pull request by ashald**
This PR provides functionality that allows mapping pillarenv values to Consul data centers.

For details see updated doc at the top of the file. Full backward compatibility.

### CODE REVIEW

File: salt/pillar/consul_pillar.py [hunk 1]
```diff
@@ -224,3 +269,50 @@ def get_conn(opts, profile):
...
+    log.debug(
+        'Selecting DC based on pillarenv using {num} pattern(s)'.format(
+            num=len(mappings)
+        )
+    )
+    log.debug('Pillarenv set to {env!r}'.format(env=pillarenv))
+
+    # sort in reverse based on pattern length
+    # but use alphabetic order within groups of patterns of same length
+    sorted_mappings = sorted(mappings, key=lambda m: (-len(m[0]), m[0]))
```
Comment (Ashald) [inline salt/pillar/consul_pillar.py]: `lambda m: (-len(m[0]), m[0])` a bit ugly but Python 2/3 compatible. :)

**Comment by Ashald** [comment]
@techhat @jfindlay @cachedout I'm sorry for disturbance, but can you please take a look at this PR? Thanks

**Comment by cachedout** [comment]
Hrm. I see where you're going with this, but this direction concerns me. This seems far too specialized for a generic pillar. I'm very hesitant about including this sort of specialized functionality into one pillar when it's not available in others and it isn't made for the general use case.

Is there something about Consul specifically that I'm not understanding which makes this sort of specialization a more natural fit? I'm totally open to this simply being a lack of knowledge on my part. :]

Let me know what you think. Thanks, @Ashald

**Comment by thatch45** [comment]
Yes, this seems to specialized, it seems like it is just making assumptions based on a specific pillar value. I will echo @cachedout in asking if there is a specific facility in Consul that this is interacting with

**Comment by Ashald** [comment]
Consul has a built-in KV storage along with multi-datacenter support. The thing is, KV namespaces are not shared across different datacenters. So, for instance, if you have key `X` in datacenter `us-east` you won't see it in datacenter `us-west`. Or you can have different values under the same key in different datacenters.

So bottom line is that KV are not replicated across datacenters but rather isolated in them. IDK about other pillars, but I see this feature as a perfect fit for integration with `pillarenv`.

**Comment by Ashald** [comment]
P.S.: Regarding:

> based on a specific pillar value
Are you talking about `pillarenv`? Correct me if I'm wrong, but isn't it's one of base features of Salt? I mean, using `saltenv` and `pillarenv` to control files and pillars environments. I though it's *the* [only] value you should be looking at when you resolving your pillars...

**Comment by thatch45** [comment]
Thanks for the clarity, I am going to give this a +1 as this makes a lot more sense.

**Comment by Ashald** [comment]
Thanks!

**Pull request merged**